common/appointment/get.ts

```diff
@@ -0,0 +1,103 @@
+import { Lecture as Appointment } from '../../graphql/generated';
+import { prisma } from '../prisma';
+import { User } from '../user';
+
+export const getAppointmentsForUser = async (
+    user: User,
+    take?: number,
+    skip?: number,
+    cursor?: number,
+    direction?: 'last' | 'next'
+): Promise<Appointment[]> => {
+    let appointments;
+
+    if (!direction && !cursor) {
+        return await getAppointmentsByIdList(user.userID, take, skip);
+    }
+
+    if (!direction || !cursor) {
+        throw Error('Cursor or direction not specified for cursor based pagination');
+    }
+
+    if (direction === 'next' && cursor) {
+        appointments = await getNextAppointmentsByIdList(user.userID, take, cursor);
+    }
+
+    if (direction === 'last' && cursor) {
+        appointments = await getLastAppointmentsByIdList(user.userID, take, cursor);
+    }
+
+    return appointments;
+};
+
+const getNextAppointmentsByIdList = async (userId: User['userID'], take: number, cursor: number) => {
+    return await prisma.lecture.findMany({
+        where: {
+            isCanceled: false,
+            OR: [
+                {
+                    participantIds: {
+                        has: userId,
+                    },
+                },
+                {
+                    organizerIds: {
+                        has: userId,
+                    },
+                },
+            ],
+        },
+        orderBy: [{ start: 'asc' }],
+        take,
+        skip: 1, // Skipping the cursor object
+        cursor: { id: cursor },
+    });
+};
+const getLastAppointmentsByIdList = async (userId: User['userID'], take: number, cursor: number) => {
+    const appointments = await prisma.lecture.findMany({
+        where: {
+            isCanceled: false,
+            OR: [
+                {
+                    participantIds: {
+                        has: userId,
+                    },
+                },
+                {
+                    organizerIds: {
+                        has: userId,
+                    },
+                },
+            ],
+        },
+        orderBy: [{ start: 'desc' }],
+        take,
+        skip: 1, // Skipping the cursor object
+        cursor: { id: cursor },
+    });
+    return appointments.reverse();
+};
+
+const getAppointmentsByIdList = async (userId: User['userID'], take, skip): Promise<Appointment[]> => {
+    return await prisma.lecture.findMany({
+        where: {
+            isCanceled: false,
+            start: { gte: new Date().toISOString() },
```

Comment:
Why do we care about the start date here, but not in the other queries?

Reply:
I'm not sure if I get the question correctly. 
Because this query is meant to give us `x = take` appointments from now and the other ones give us `x`  appointments from a specific cursor in a specific direction. So we're [paginating cursor based](https://www.prisma.io/docs/concepts/components/prisma-client/pagination#cursor-based-pagination) to be able to scroll upwards and downwards infinitely in a scrollable list.

Reply:
Tbh I never worked with cursors, so I'm somewhat confused by the function naming, but the concept looks interesting. 

Could you give me an example how the three functions will be used together?

Reply:
Sure:
Imagine a list of tiles (a list of appointments is a pretty good use case). First, I want the upcoming appointments that start in the future from "now". So we query 10 items from now. When I reach the bottom of the list, the next 5 appointments after the last appointment (`cursor: lastId, direction: 'next', take: 5`) are queried and merged together with the current list. But this is not the point, because this can be achieved with a offset pagination. But if I now scroll to the top of the list to get to the past appointments, I use the same query again, but with different parameters (`cursor: firstIdInList, direction: 'last', take: 5` ), and I can merge the queryResult with the current list again. 

Did it make it clearer?

Reply:
> I'm somewhat confused by the function naming [...]

The naming was a copy and paste artifact 😄 I'll refactor that.